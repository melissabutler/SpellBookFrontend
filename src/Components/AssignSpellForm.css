.AssignSpellForm {
    text-align: center;
}

.AssignSpellForm-form {
    margin: 20px;
}

select {
    margin: 1rem;
    width: 50vw;
    padding: 1rem;
    border-radius: 10px;
    font-size: 20px;
}

option {
    padding: 1rem;
}
input {
    font-size: 20px;
    text-align: center;
    margin: 1rem;
    width: 50vw;
    padding: 1rem;
    
    border-radius: 10px;
}
button {
    margin: 20px;
    float: none;
}