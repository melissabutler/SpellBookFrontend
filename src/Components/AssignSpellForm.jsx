import React, { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import "./Form.css"

import { Container, Row, Col } from "react-bootstrap"

import CurrentUserContext from "../currentUserContext";
import useFields from "../Hooks/useFields"

const AssignSpellForm = ({assignSpell, spellIdx, getUser}) => {
    let navigate = useNavigate();
    const [characters, setCharacters] = useState([])

    useEffect( () => {
        async function getCharacters() {
            let data = await getUser();
            setCharacters([...data.user.characters])
        }
        getCharacters();
    }, [])

    let [formData, handleChange, resetFormData] = useFields({
        char_id: undefined,
    })

    const handleSubmit = async e => {
        e.preventDefault();
        if(!formData.char_id){
            alert("Please choose a character to assign the spell.")
        }
        let assignedCharacter = {
            "id": parseInt(formData.char_id),
        } 
        
        await assignSpell(spellIdx, assignedCharacter.id)
        navigate(`/characters/${assignedCharacter.id}`)
    }

    return (
        <Container className="Form">
            <form className="Form-form" onSubmit={handleSubmit}>
                <Row className="Form-label">
                    <Col></Col>
                    <Col>
                     <label htmlFor="id">Assign spell to character:</label>
                    </Col>
                   <Col></Col>
                </Row>
                <Row>
                    <Col></Col>
                    <Col>
                  <select id="char_id" value={formData.char_id} name="char_id" onChange={handleChange}>
                    <option key="placeholder" value="">Select character</option>
                    {characters.map(character => (
                        <option key={character.id} value={character.id}>{character.char_name}</option>
                    ))}
                </select>  
                </Col>
                <Col></Col>
                </Row>
                <Row>
                    <Col></Col>
                    <Col>
                     <button> Submit</button>
                    </Col>
                    <Col></Col>
                   
                </Row>
                
            </form>

        </Container>
    )
}

export default AssignSpellForm;
