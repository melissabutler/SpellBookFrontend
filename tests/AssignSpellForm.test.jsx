import React from "react";
import { describe, it, expect, vi } from "vitest";
import { render, screen} from '@testing-library/react'
import AssignSpellForm from "../src/Components/AssignSpellForm";

describe("AssignSpellForm", function() {
    vi.mock('react', async (importOriginal) => {
                const react = await importOriginal('react');
                return {
                    ...react,
                    useContext: () => ({
                        username: 'u1',
                        email: 'email@email.com',
                        isAdmin: false,
                        characters: [{ char_class: "cleric", char_name: "Beebo", id: 1, lvl: 3 },
                        { char_class: "cleric", char_name: "Melba", id: 2, lvl: 3 }
                        ]
                    }),
                    useRef: () => { target: null }
            }})

        vi.mock('react-router-dom', async () => {
            const mod = await vi.importActual('react-router-dom');
            return {
                ...mod, 
                useParams: () => ({
                    idx: "acid-arrow"
                }),
                useNavigate: () => "/"
            }
        })
    it('renders the AssignSpellForm component', async () => {
        render(
            <AssignSpellForm />)
        expect(screen.getByText("Assign spell to character:")).toBeInTheDocument();
      screen.debug()
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<AssignSpellForm />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("has user characters in Select character list", async () => {
        render(<AssignSpellForm />)
        expect(screen.getByText("Beebo")).toBeInTheDocument();
        expect(screen.getByText('Melba')).toBeInTheDocument();
    })



})
   