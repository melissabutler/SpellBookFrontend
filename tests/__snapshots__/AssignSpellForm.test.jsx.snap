// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AssignSpellForm > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Form container"
  >
    <form
      class="Form-form"
    >
      <div
        class="Form-label row"
      >
        <div
          class="col"
        />
        <div
          class="col"
        >
          <label
            for="id"
          >
            Assign spell to character:
          </label>
        </div>
        <div
          class="col"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        />
        <div
          class="col"
        >
          <select
            id="char_id"
            name="char_id"
          >
            <option
              value=""
            >
              Select character
            </option>
            <option
              value="1"
            >
              Beebo
            </option>
            <option
              value="2"
            >
              Melba
            </option>
          </select>
        </div>
        <div
          class="col"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        />
        <div
          class="col"
        >
          <button>
             Submit
          </button>
        </div>
        <div
          class="col"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;
